File: src/main/scala/com/leetcode/binarysearch/BinarySearch.scala
package com.leetcode.binarysearch

import scala.annotation.tailrec

object BinarySearch extends App {
  def binarySearch(ds: Array[Double], key: Double): Option[Int] = {
    @tailrec
    def go(lo: Int, hi: Int): Option[Int] = {
      if(lo > hi) None
      else {
        val mid: Int = lo + (hi -lo)/2
        ds(mid) match {
          case mv if (mv == key) => Some(mid)
          case mv if (mv <= key) => go(mid + 1, hi)
          case _ => go(lo, mid - 1)
        }
      }
    }
    go(0, ds.size - 1)
  }

}

File: src/main/java/com/leetcode/binarysearch/BinarySearch.java
package com.leetcode.binarysearch;

public class BinarySearch {
    public int binarySearch(int[] nums, int target) {
        if(nums == null || nums.length == 0) return -1;

        int left  =0, right = nums.length - 1;
        while(left <= right) {
            //prevent (left + right) overflow
            int mid = left + (right - left) / 2;
            if(nums[mid] == target) return mid;
            else if(nums[mid] < target) left = mid + 1;
            else right = mid - 1;
        }
        return -1;
    }

    public int searchRotatedArray(int[] nums, int target) {
        int start = 0;
        int end = nums.length - 1;
        while(start <= end) {
            int mid = (start + end)/2;
            if(nums[mid] == target) return mid;
            if(nums[start] <= nums[mid]) {
                if(target < nums[mid] && target > nums[start])
                    end = mid - 1;
                else
                    start = mid + 1;
            }

            if(nums[mid] <= nums[end]) {
                if(target > nums[mid] && target <= nums[end])
                    start = mid + 1;

                else
                        end = mid - 1;
                }
            }

        return -1;
    }

    public static void main(String[] args) {
        BinarySearch objbinarySearch = new BinarySearch();
        int[] input = {10, 20, 30, 40, 50};
        int index = objbinarySearch.binarySearch( input, 30);
        System.out.println(index);
    }
}


